sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StructuredMessageDecoder.java

```diff
@@ -0,0 +1,182 @@
+package com.azure.storage.common.implementation;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.HashMap;
+import java.util.Map;
+
+public class StructuredMessageDecoder {
+    private static final int DEFAULT_MESSAGE_VERSION = 1;
+    private static final int V1_HEADER_LENGTH = 13;
+    private static final int V1_SEGMENT_HEADER_LENGTH = 10;
+    private static final int CRC64_LENGTH = 8;
+
+    private int messageVersion;
+    private int messageLength;
+    private StructuredMessageFlags flags;
+    private int numSegments;
+    private final ByteBuffer buffer;
+
+    private int messageOffset = 0;
+    private int currentSegmentNumber = 0;
+    private int currentSegmentContentLength = 0;
+    private int currentSegmentContentOffset = 0;
+
+    private long messageCrc64 = 0;
+    private long segmentCrc64 = 0;
+    private final Map<Integer, Long> segmentCrcs = new HashMap<>();
+
+    public StructuredMessageDecoder(ByteBuffer inputBuffer) {
```

Comment:
In terms of the constructor, I'm thinking instead of accepting an input byte buffer, we'd accept the expected content length of the individual download and the `decode()` method would accept a byte buffer containing a part of the entire encoded structured message. The main reason that I'm suggesting this is that I suspect in order to integrate this with the client, we will be provided a flux of byte buffers so the content of the entire encoded structured message will be broken across multiple buffers. I imagine it will be similar in integration style to the client side encryption Decryptor class (note that it accepts a `Flux<ByteBuffer>` but we'll just want to accept `ByteBuffer` for now so that Reactor integration is decoupled from logic to do the actual decoding)